File: internal/tools/capabilities.go
package tools

import (
	"context"
	"encoding/json"
	"fmt"
	"sync"
	"time"

	"github.com/Howthemeaning/gnmi-mcp-server/internal/config"
	"github.com/Howthemeaning/gnmi-mcp-server/internal/gnmi"
	"github.com/modelcontextprotocol/go-sdk/mcp"
)

type CapInput struct {
	ConnParams
}

type capCacheEntry struct {
	data   json.RawMessage
	expiry time.Time
}

type capCache struct {
	mu sync.Mutex
	m  map[string]capCacheEntry
}

func newCapCache() *capCache { return &capCache{m: map[string]capCacheEntry{}} }

func (c *capCache) get(key string) (json.RawMessage, bool) {
	c.mu.Lock()
	defer c.mu.Unlock()
	e, ok := c.m[key]
	if !ok || time.Now().After(e.expiry) {
		return nil, false
	}
	return e.data, true
}

func (c *capCache) set(key string, data json.RawMessage) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.m[key] = capCacheEntry{data: data, expiry: time.Now().Add(5 * time.Minute)}
}

func RegisterCapabilities(server *mcp.Server, client gnmi.GnmiClient, cfg *config.AppConfig) {
	cache := newCapCache()
	mcp.AddTool(server, &mcp.Tool{
		Name:        "gnmi_capabilities",
		Description: "Query a gNMI device's version, supported YANG models and encodings.",
	}, func(ctx context.Context, _ *mcp.CallToolRequest, in CapInput) (*mcp.CallToolResult, any, error) {
		dev, err := in.apply(cfg)
		if err != nil {
			return textResult(err.Error(), true), nil, nil
		}
		key := fmt.Sprintf("%s|insecure=%t|skipverify=%t", dev.Address, dev.Insecure, dev.SkipVerify)
		if cached, ok := cache.get(key); ok {
			return textResult(string(cached), false), nil, nil
		}
		resp, err := client.Capabilities(ctx, dev)
		if err != nil {
			return textResult(err.Error(), true), nil, nil
		}
		cache.set(key, resp)
		return textResult(string(resp), false), nil, nil
	})
}

File: src/gnmi_mcp_server/tools/capabilities.py
"""MCP tool: gnmi_capabilities -- query gNMI device capabilities."""

import time
import logging

logger = logging.getLogger(__name__)

_capabilities_cache: dict = {}


def register_capabilities_tool(server, gnmic_client, config):
    """Register the gnmi_capabilities tool."""

    @server.tool()
    async def gnmi_capabilities(
        target: str = "",
        address: str = "",
        insecure: bool = False,
        tls_ca: str = "",
        tls_cert: str = "",
        tls_key: str = "",
    ) -> str:
        """Query gNMI device capabilities (version, supported models, encodings)."""
        from gnmi_mcp_server.tools._common import resolve_device, build_extra_args

        try:
            device = resolve_device(config, target, address)
        except ValueError as e:
            return str(e)

        # 对 adhoc 设备（无预定义 name）用 address 区分缓存
        name_key = device.address if device.name == "__adhoc__" else device.name
        cache_key = (name_key, "JSON_IETF", insecure)
        if cache_key in _capabilities_cache:
            cached_result, expiry = _capabilities_cache[cache_key]
            if time.time() < expiry:
                return cached_result

        extra_args = build_extra_args(device, insecure, tls_ca, tls_cert, tls_key)
        result = await gnmic_client.run("capabilities", extra_args, device)

        if result.is_success:
            output = result.stdout
            _capabilities_cache[cache_key] = (output, time.time() + 300)
            return output
        else:
            return f"Error: {result.error_message}"
